test/test_device_batch_copy.cu

```diff
@@ -0,0 +1,427 @@
+
+  std::mt19937 rng(0);
+  std::uniform_int_distribution<std::size_t> size_dist(1, 1000000);
+  for (std::size_t i = 0; i < num_rnd_buffer_range_tests; i++)
+  {
+    auto range_begin = size_dist(rng);
+    auto range_end   = size_dist(rng);
+    if (range_begin > range_end)
+    {
+      std::swap(range_begin, range_end);
+    }
+    buffer_size_ranges.push_back({range_begin, range_end});
+  }
+
+  for (const auto &buffer_size_range : buffer_size_ranges)
+  {
+    BufferSizeT min_buffer_size =
+      static_cast<BufferSizeT>(CUB_ROUND_UP_NEAREST(buffer_size_range.first, sizeof(AtomicCopyT)));
+    BufferSizeT max_buffer_size =
+      static_cast<BufferSizeT>(CUB_ROUND_UP_NEAREST(buffer_size_range.second,
+                                                    static_cast<BufferSizeT>(sizeof(AtomicCopyT))));
+    double average_buffer_size = (min_buffer_size + max_buffer_size) / 2.0;
+    BufferOffsetT target_num_buffers =
+      static_cast<BufferOffsetT>(target_copy_size / average_buffer_size);
+
+    // Run tests with input buffer being consecutive and output buffers being consecutive
+    RunTest<AtomicCopyT, BufferOffsetT, BufferSizeT, ByteOffsetT>(target_num_buffers,
+                                                                  min_buffer_size,
+                                                                  max_buffer_size,
+                                                                  TestDataGen::CONSECUTIVE);
+
+    // Run tests with input buffer being randomly shuffled and output buffers being randomly
+    // shuffled
+    RunTest<AtomicCopyT, BufferOffsetT, BufferSizeT, ByteOffsetT>(target_num_buffers,
+                                                                  min_buffer_size,
+                                                                  max_buffer_size,
+                                                                  TestDataGen::RANDOM);
+  }
+
+  //---------------------------------------------------------------------
+  // DeviceCopy::Batched test with 64-bit offsets
+  //---------------------------------------------------------------------
+  using ByteOffset64T = uint64_t;
+  using BufferSize64T = uint64_t;
+  ByteOffset64T large_target_copy_size =
+    static_cast<ByteOffset64T>(std::numeric_limits<uint32_t>::max()) + (128ULL * 1024ULL * 1024ULL);
+  // Make sure min_buffer_size is in fact smaller than max buffer size
+  constexpr BufferOffsetT single_buffer = 1;
+
+  // Run tests with input buffer being consecutive and output buffers being consecutive
+  RunTest<AtomicCopyT, BufferOffsetT, BufferSize64T, ByteOffset64T>(single_buffer,
+                                                                    large_target_copy_size,
+                                                                    large_target_copy_size,
+                                                                    TestDataGen::CONSECUTIVE);
```

Comment:
I think this has quite some heavy memory requirements with a 64-bit `AtomicCopyT`. Maybe we can lower the memory requirements when checking 64-bit indexes by using a 8-bit`AtomicCopyT`?